### DISCUSSION TITLE
Any chance of support for embedded platforms?

### DISCUSSION THREAD

**Discussion by J3RN** [GetFirefly/firefly#687] [Q&A] [answered]
Hello all! :wave: I'm interested as to whether it would be possible to port Lumen to MCUs such as Tensilica Xtensa or Arm Cortex-M SoCs.

### Rationale

I'm a fan of [the Nerves project](https://hexdocs.pm/nerves/), but one if its weaknesses IMO is that it requires at least a (relatively beefy) Arm Cortex-A series chip in order to run. This is likely in part because it runs atop a buildroot Linux environment.

Does the AoT compilation aspect (or any other aspects) of Lumen make it more amenable to being able to run on resource-constrained MCUs?

**Comment by KronicDeth** [comment, marked as answer]
It was always part of our plan to target Lumen to embedded systems and we've talked about it in our conference presentations before.  Lumen being AoT is there to help with the resource constraints of both browsers and embedded systems.  On x86_64 it's more useful for competing with single binary executable commandline/agent stuff like vs Go or Rust.

Without a funding/development partner for a specific embedded system, it would likely be a priority behind WASM and desktop/server (x86_64 and now M1 to support new Macs).  x86_64/M1 is a necessary target to allow us to test and develop Lumen.  As you can see from the CI runs, we still need that working all the way to get OTP to compile before we focus on another hardware target.
